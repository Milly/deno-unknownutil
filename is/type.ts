/**
 * A type predicate function.
 */
export type Predicate<T> = (x: unknown) => x is T;

/**
 * A type predicated by Predicate<T>.
 *
 * ```ts
 * import { is, type PredicateType } from "https://deno.land/x/unknownutil@$MODULE_VERSION/mod.ts";
 *
 * const isPerson = is.ObjectOf({
 *   name: is.String,
 *   age: is.Number,
 *   address: is.OptionalOf(is.String),
 * });
 *
 * type Person = PredicateType<typeof isPerson>;
 * // Above is equivalent to the following type
 * // type Person = {
 * //   name: string;
 * //   age: number;
 * //   address: string | undefined;
 * // };
 */
export type PredicateType<P> = P extends Predicate<infer T> ? T : never;
