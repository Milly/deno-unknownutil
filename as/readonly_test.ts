import { assertSnapshot } from "@std/testing/snapshot";
import { assertType } from "@std/testing/types";
import type { Equal } from "../_testutil.ts";
import { is } from "../is/mod.ts";
import { asReadonly, asUnreadonly } from "./readonly.ts";

Deno.test("asReadonly<T>", async (t) => {
  await t.step("returns properly named function", async (t) => {
    await assertSnapshot(t, asReadonly(is.Number).name);
    // Nesting does nothing
    await assertSnapshot(t, asReadonly(asReadonly(is.Number)).name);
  });
  await t.step("returns proper type predicate", () => {
    const a: unknown = undefined;
    if (is.ObjectOf({ a: asReadonly(is.Number) })(a)) {
      assertType<Equal<typeof a, { readonly a: number }>>(true);
    }
  });
});

Deno.test("asUnreadonly<T>", async (t) => {
  await t.step("returns properly named function", async (t) => {
    await assertSnapshot(t, asUnreadonly(asReadonly(is.Number)).name);
    // Non optional does nothing
    await assertSnapshot(t, asUnreadonly(is.Number).name);
    // Nesting does nothing
    await assertSnapshot(
      t,
      asUnreadonly(asUnreadonly(asReadonly(is.Number))).name,
    );
  });
  await t.step("returns proper type predicate", () => {
    const a: unknown = undefined;
    if (is.ObjectOf({ a: asUnreadonly(asReadonly(is.Number)) })(a)) {
      assertType<Equal<typeof a, { a: number }>>(true);
    }
  });
});
