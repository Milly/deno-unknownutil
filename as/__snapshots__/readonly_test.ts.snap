export const snapshot = {};

snapshot[`asReadonly<T> > returns properly named function 1`] = `"asReadonly(isNumber)"`;

snapshot[`asReadonly<T> > returns properly named function 2`] = `"asReadonly(isNumber)"`;

snapshot[`asUnreadonly<T> > returns properly named function 1`] = `"isNumber"`;

snapshot[`asUnreadonly<T> > returns properly named function 2`] = `"isNumber"`;

snapshot[`asUnreadonly<T> > returns properly named function 3`] = `"isNumber"`;
